modules/tts/tts-utils/tts-common/src/main/java/org/daisy/pipeline/tts/ExponentialBackoffScheduler.java

```diff
@@ -0,0 +1,50 @@
+package org.daisy.pipeline.tts;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+public class ExponentialBackoffScheduler<RequestType> implements RequestScheduler<RequestType> {
+	
+	private int n = 0;
```

Comment:
I find it not so intuitive that this number is never reset. It probably doesn't matter because for every invocation of `px:ssml-to-audio` a new `GoogleRestTTSEngine` is created with a new `ExponentialBackoffScheduler`, but it is not really clear that the class is supposed to be used like that. Maybe the documentation just needs some improvement.

In general, I think it is not super clear how the `RequestScheduler` interface is supposed to be used. More specifically, in what order are the methods supposed to be called. Would it be an idea to extract some of the logic from `GoogleRestTTSEngine` into the `org.daisy.pipeline.tts` package (tts-common) to make the interface more useful?

Reply:
You are right, i missed that >< 
Indeed the counter is meant to be reset if a request is successfull after the last delay.

I realise my scheduler interface needs some improvements, like some intel about retries and/or more functions like cancel and resume to respectively cancel a request or remove/reset the delay on a specified (or all) request(s)
I will update the interface and the javadoc to better reflect the idea of what the scheduler could do.

For the RequestScheduler, it is meant to delay the execution of requests based on constraints to be defined (like if an error is raised on a request or if quotas are met).

For the logic behind the TTS for rest API, i imagine some of the logic can be shared, as the main changes are the parameters and the headers of each request, but i'm not sure yet.
Maybe some kind of `RestRequest` interface could be defined and use in a `RestTTSEngine` with the `RequestScheduler` (if this is what you are thinking of)

Reply:
I was thinking of something more low level than that. Nothing to do with TTS or REST.

What I had in mind was just a way to let a scheduler execute certain actions. Basically you feed it "runnables" that may throw an exception when the action failed, in which case they are automatically rescheduled.

A very quick sketch to make the idea clear:

```java
interface Action {
    public void apply() throws RecoverableError, FatalError;
}

void process(RequestScheduler<Action>, Action) throws FatalError {
    ...
}

ExponentialBackoffScheduler<Action> scheduler = new ExponentialBackoffScheduler<>();

GoogleRestRequest speechRequest = ...;

process(
    scheduler,
    () -> {
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(speechRequest.send(), "utf-8"));
            StringBuilder response = new StringBuilder();
            String inputLine;
            while ((inputLine = br.readLine()) != null)
                response.append(inputLine.trim());
            br.close();
            byte[] decodedBytes = Base64.getDecoder().decode(response.toString().substring(18, response.length()-2));
            AudioBuffer b = bufferAllocator.allocateBuffer(decodedBytes.length);
            b.data = decodedBytes;
            result.add(b);
        } catch (Throwable e1) {
            try {
                if (request.getConnection().getResponseCode() == 429)
                    throw new RecoverableError();
                else {
                    SoundUtil.cancelFootPrint(result, bufferAllocator);
                    StringWriter sw = new StringWriter();
                    e1.printStackTrace(new PrintWriter(sw));
                    throw new FatalError(new SynthesisException(e1));
                }
            } catch (Exception e2) {
                SoundUtil.cancelFootPrint(result, bufferAllocator);
                StringWriter sw = new StringWriter();
                e2.printStackTrace(new PrintWriter(sw));
                throw new FatalError(new SynthesisException(e2));
            }
        }
    }
);
```

Reply:
Ok i'll work on that